synapse/lib/httpapi.py

```diff
@@ -0,0 +1,498 @@
+
+    def sendRestRetn(self, valu):
+        return self.write({'status': 'ok', 'result': valu})
+
+    def loadJsonMesg(self, byts):
+        try:
+            return json.loads(byts)
+        except Exception as e:
+            self.sendRestErr('BadJson', 'Invalid JSON content.')
+            return None
+
+    def sendAuthReqired(self):
+        self.set_header('WWW-Authenticate', 'Basic realm=synapse')
+        self.set_status(401)
+        self.sendRestErr('NotAuthenticated', 'The session is not logged in.')
+
+    async def reqAuthUser(self):
+        if await self.authenticated():
+            return True
+        self.sendAuthReqired()
+        return False
+
+    async def reqAuthAdmin(self):
+
+        if self.cell.insecure:
+            return True
+
+        user = await self.user()
+        if user is None:
+            self.sendAuthReqired()
+            return False
+
+        if not user.admin:
+            self.sendRestErr('AuthDeny', f'User {user.iden} ({user.name}) is not an admin.')
+            return False
+
+        return True
+
+    async def sess(self, gen=True):
+
+        if self._web_sess is None:
+
+            iden = self.get_secure_cookie('sess')
+
+            if iden is None and not gen:
+                return None
+
+            if iden is None:
+                iden = s_common.guid()
+                opts = {'expires_days': 14, 'secure': True, 'httponly': True}
+                self.set_secure_cookie('sess', iden, **opts)
+
+            self._web_sess = await self.cell.genHttpSess(iden)
+
+        return self._web_sess
+
+    async def user(self):
+
+        if self._web_user is not None:
+            return self._web_user
+
+        sess = await self.sess(gen=False)
+        if sess is not None:
+            return sess.user
+
+        auth = self.request.headers.get('Authorization')
+        if auth is None:
+            return None
+
+        if not auth.startswith('Basic '):
+            return None
+
+        _, blob = auth.split(None, 1)
+
+        try:
+            text = base64.b64decode(blob).decode('utf8')
+            name, passwd = text.split(':', 1)
+        except Exception:
+            logger.exception('invalid basic auth header')
+            return None
+
+        user = self.cell.auth.getUserByName(name)
+        if user is None:
+            return None
+
+        if user.locked:
+            return None
+
+        if not user.tryPasswd(passwd):
+            return None
+
+        self._web_user = user
+        return user
+
+    async def username(self):
+        user = await self.user()
+        if user is not None:
+            return user.name
+
+    async def authenticated(self):
+        if self.cell.insecure:
+            return True
+        return await self.user() is not None
+
+class WebSocket(HandlerBase, t_websocket.WebSocketHandler):
+    pass
+
+class Handler(HandlerBase, t_web.RequestHandler):
+    pass
+
+class StormNodesV1(Handler):
```

Comment:
given that this and StormV1 are both cortex specific, it feels a bit odd to have them live here; unless we indend to define *all* core httpapi handlers in this file.

Reply:
Agreed.  We may eventually want to create some kind of package level breakdown for http API elements for different service types.  I think once we implement a few more, it'll be worth breaking them out.  I think they're probably fine for now though.